File: backend/api/application/use_cases/contact_use_case.py
"""Contact use cases - application logic for contact form operations."""

from api.application.interfaces.email_service import IEmailService
from api.domain.exceptions import EmailServiceException
from api.domain.value_objects import ContactMessage, Email


class ContactUseCase:
    """Use cases for contact form operations."""

    def __init__(
        self,
        email_service: IEmailService,
        admin_email: Email,
        site_name: str = "NeoTemplate",
    ) -> None:
        """
        Initialize contact use case.

        Args:
            email_service: Service for sending emails
            admin_email: Admin email to receive contact messages
            site_name: Name of the site (for email subject)
        """
        self._email_service = email_service
        self._admin_email = admin_email
        self._site_name = site_name

    def send_contact_message(self, contact_message: ContactMessage) -> None:
        """
        Send a contact message to admin.

        Args:
            contact_message: Contact message value object

        Raises:
            EmailServiceException: If email sending fails
        """
        subject = f"Message from a user on {self._site_name}"
        message_body = f"{contact_message.email.value} vous a écrit:\n\n{contact_message.message}"

        try:
            self._email_service.send_email(
                subject=subject,
                message=message_body,
                from_email=contact_message.email,
                recipient_list=[self._admin_email],
            )
        except Exception as e:
            raise EmailServiceException(
                f"Failed to send contact message: {str(e)}"
            ) from e

File: frontend/src/application/use-cases/contact-use-case.ts
/**
 * Contact use cases - application logic for contact form operations.
 */

import { EmailServiceException, ValidationException } from '@/src/domain/exceptions';
import { ContactMessage, Email } from '@/src/domain/value-objects';
import { IApiClient, ContactApiResponse } from '@/src/domain/interfaces/api-client.interface';

/**
 * Contact form submission result.
 */
export interface ContactResult {
  success: boolean;
  message: string;
}

/**
 * Contact use cases.
 */
export class ContactUseCase {
  constructor(private readonly apiClient: IApiClient) {}

  /**
   * Send a contact message.
   *
   * @param email - Sender email
   * @param message - Contact message
   * @returns Contact result
   * @throws {ValidationException} If message data is invalid
   * @throws {EmailServiceException} If message cannot be sent
   */
  async sendContactMessage(
    email: string,
    message: string
  ): Promise<ContactResult> {
    try {
      // Create domain value objects with validation
      const emailValue = new Email(email);
      const contactMessage = new ContactMessage(emailValue, message);

      // Convert to FormData for API (matches backend expectations)
      const formData = new FormData();
      formData.append('email', contactMessage.email.value);
      formData.append('message', contactMessage.message);

      await this.apiClient.post<ContactApiResponse>(
        '/api/contact/',
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        }
      );

      return {
        success: true,
        message: 'Your message has been sent, thank you.',
      };
    } catch (error: unknown) {
      if (error instanceof ValidationException) {
        throw error;
      }

      // Handle API validation errors
      if (
        error &&
        typeof error === 'object' &&
        'errors' in error &&
        error.errors
      ) {
        throw new ValidationException(
          'Contact form validation failed',
          error.errors as Record<string, string[]>
        );
      }

      // Handle email service errors
      if (
        error &&
        typeof error === 'object' &&
        'code' in error &&
        error.code === 'email_service_error'
      ) {
        throw new EmailServiceException(
          'An error occurred. The message could not be sent.'
        );
      }

      throw new EmailServiceException(
        `Failed to send contact message: ${
          error instanceof Error ? error.message : 'Unknown error'
        }`
      );
    }
  }
}

